src/daft-logical-plan/src/optimization/rules/push_down_join_predicate.rs

```diff
@@ -0,0 +1,205 @@
+use common_error::DaftResult;
+use common_treenode::{Transformed, TreeNode};
+use daft_core::join::{JoinSide, JoinType};
+
+use super::OptimizerRule;
+use crate::{
+    ops::{Filter, Join},
+    LogicalPlan, LogicalPlanRef,
+};
+
+/// Optimizer rule for pushing down join predicates that only need one side of the join.
+///
+/// # Example
+/// In the following query, `a.y = 1` can be pushed down:
+/// ```sql
+/// -- before
+/// SELECT * FROM a JOIN b ON a.x = b.x AND a.y = 1
+/// -- after
+/// SELECT * FROM (SELECT * FROM a WHERE a.y = 1) JOIN b ON a.x = b.x
+/// ```
+#[derive(Debug)]
+pub struct PushDownJoinPredicate {}
+
+impl PushDownJoinPredicate {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl OptimizerRule for PushDownJoinPredicate {
+    fn try_optimize(&self, plan: LogicalPlanRef) -> DaftResult<Transformed<LogicalPlanRef>> {
+        plan.transform(|node| {
+            if let LogicalPlan::Join(Join {
+                left,
+                right,
+                on,
+                join_type,
+                join_strategy,
+                ..
+            }) = node.as_ref()
+            {
+                let (can_push_down_left, can_push_down_right) = match join_type {
+                    JoinType::Inner => (true, true),
+                    JoinType::Left => (false, true),
+                    JoinType::Right => (true, false),
```

Comment:
Hm I might be misunderstanding but I don't actually think this is true.

If we have `A left join B WHERE A.x = B.x AND B.y = 1` we can't push `B.y = 1` down because `B.y = 1` removed nulls from the join.

Reply:
`A left join B where B.x = 1` does remove nulls, but it's actually quite different from `A left join B on B.x = 1` (note the "on"). The latter does not remove nulls and is what we are pushing down here

Reply:
Ah! I completely misread TPCH Q13 too it seems. Ok gotcha